### PR TITLE
added doRefund

### PR DESCRIPTION
**Pull request by Plikaplouk**
todo : the refund action should be in a static variable

### CODE REVIEW

File: Payline/Payline.php [hunk 1]
```diff
@@ -156,4 +156,30 @@ public function verifyWebTransaction($paymentToken)
 
         return $paylineResult;
     }
+    
+    public function doRefund($paymentToken, $comment = '', $sequenceNumber = 0) {
```
Comment (lolautruche) [inline Payline/Payline.php]: CS: `{` must be on a separate line, with no blank line below

File: Payline/Payline.php [hunk 2]
```diff
@@ -156,4 +156,30 @@ public function verifyWebTransaction($paymentToken)
...
+        // first, we get the payment details
+        $paymentDetails = $this->paylineSDK->getWebPaymentDetails(['token' => $paymentToken]);
+
+        // add the private datas
+        foreach ($paymentDetails['privateDataList']['privateData'] as $key => $value) {
+            $this->paylineSDK->addPrivateData(['key' => $value['key'], 'value' => $value['value']]);
+        }
+
+        // refund action is 421
+        $paymentDetails['payment']['action'] = '421';
```
Comment (lolautruche) [inline Payline/Payline.php]: Please use a class constant

Comment (lolautruche) [inline Payline/Payline.php]: Class constant should preferably be on `WebGatewayInterface`

File: Payline/Payline.php [hunk 3]
```diff
@@ -156,4 +156,30 @@ public function verifyWebTransaction($paymentToken)
...
+
+        $params = array(
+            'transactionID' => $paymentDetails['transaction']['id'],
+            'payment' => $paymentDetails['payment'],
+            'comment' => $comment,
+            'sequenceNumber' => $sequenceNumber
+        );
```
Comment (lolautruche) [inline Payline/Payline.php]: Please use short array syntax

File: Payline/Payline.php (hunk 1 above)
Comment (lolautruche) [inline Payline/Payline.php]: Please also add the method in the interface, with PHPDoc

**Comment by lolautruche** [comment]
Thanks @Plikaplouk for the PR!

I added a few comments.
It would also be nice to add tests 😉

**Comment by Plikaplouk** [comment]
I read your comments and I did the change, tell me if something is missing :)

File: Payline/Payline.php [hunk 4]
```diff
@@ -156,4 +156,29 @@ public function verifyWebTransaction($paymentToken)
...
+    {
+        // first, we get the payment details
+        $paymentDetails = $this->paylineSDK->getWebPaymentDetails(['token' => $paymentToken]);
+
+        // add the private datas
+        foreach ($paymentDetails['privateDataList']['privateData'] as $key => $value) {
+            $this->paylineSDK->addPrivateData(['key' => $value['key'], 'value' => $value['value']]);
+        }
+
+        $paymentDetails['payment']['action'] = WebGatewayInterface::CODE_ACTION_DOREFUND;
```
Comment (lolautruche) [inline Payline/Payline.php]: Constant doesn't seem to be defined in the interface...

**Comment by lolautruche** [comment]
OK, apart from the missing constant, we're only missing new tests.
Almost ready to merge :-)

**Comment by lolautruche** [comment]
Thanks for the change!
As for the tests, if you don't know how to do, just allow me to edit your PR. You should have a button somewhere on the PR page.

**Comment by Plikaplouk** [comment]
I haven't had time to do the test yet.
If you want to do it, you are already allowed to (the checkbox 'Allow edits from mainteners' is checked.
Thank you !

**Comment by lolautruche** [comment]
Awesome !

I couldn't find the time to do it myself, sorry for that.
I'm merging this in right away 😄

**Comment by lolautruche** [comment]
Just released v1.1.0

**Pull request merged**